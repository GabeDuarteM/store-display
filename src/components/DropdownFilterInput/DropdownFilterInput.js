import React from 'react'

import PropTypes from 'prop-types'
import cn from 'classnames'
import injectSheet from 'react-jss'

import downArrow from '../../assets/down-arrow.svg'
import SvgIcon from '../SvgIcon'

export const styles = {
  dropdownFilter: {
    WebkitAppearance: 'none !important',
    boxSizing: 'border-box',
    height: 36,
    paddingLeft: 15,
    width: ({ width }) => width,
    alignItems: 'center',
    display: 'flex',
    fontSize: 12,
    border: ({ isOpen }) => `1px solid #${isOpen ? '6e8be3' : 'd7d8db'}`,
    borderRadius: 3,
    color: '#585858',
    backgroundColor: 'white',
    '&:active': {
      border: '1px solid #6e8be3',
    },
  },
  downArrow: {
    marginLeft: 'auto',
    paddingRight: 16,
    '&.isOpen': {
      transform: 'rotateX(180deg)',
    },
  },
}

const DropdownFilterInput = ({
  classes,
  getButtonProps,
  handleChangeMenuIsOpen,
  selectedItem,
  isOpen,
}) => (
  <div
    role="button"
    {...getButtonProps({
      onClick: () => {
        handleChangeMenuIsOpen(!isOpen)
      },
      className: classes.dropdownFilter,
    })}
  >
    {selectedItem.text}
    <SvgIcon
      className={cn(classes.downArrow, { isOpen })}
      width={9}
      color="#9B9B9B"
      src={downArrow}
    />
  </div>
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** Function to get the props from Downshift */
  getButtonProps: PropTypes.func.isRequired,
  /** Property that define if the item list is opened or not */
  isOpen: PropTypes.bool.isRequired,
  /** Gets called when the user clicks the input. It used to open the menu with the items. */
  handleChangeMenuIsOpen: PropTypes.func.isRequired,
  /** The current selected item to be displayed on the input */
  selectedItem: PropTypes.shape({
    id: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
  }).isRequired,
  /** The width of the component */
  width: PropTypes.number.isRequired,
}

DropdownFilterInput.propTypes = propTypes

export default injectSheet(styles)(DropdownFilterInput)
