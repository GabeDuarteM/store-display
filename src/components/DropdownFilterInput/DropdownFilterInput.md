```js
<DropdownFilterInput
  getButtonProps={customProps => customProps}
  handleChangeMenuIsOpen={() => {}}
  isOpen={false}
  selectedItem={{ id: 'price-asc', text: 'Price: low to high' }}
  width={200}
/>
```
