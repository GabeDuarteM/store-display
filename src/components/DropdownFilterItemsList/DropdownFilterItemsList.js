import React from 'react'

import PropTypes from 'prop-types'
import injectSheet from 'react-jss'

import checkMark from '../../assets/check-mark.svg'
import SvgIcon from '../SvgIcon'

export const styles = {
  itemList: {
    boxSizing: 'border-box',
    width: ({ width }) => width,
    position: 'absolute',
    marginTop: 4,
    cursor: 'pointer',
    border: '1px solid #e7e8ec',
    boxShadow: '0px 2px 6px rgba(115,115,115,0.12)',
    borderRadius: 3,
    backgroundColor: 'white',
    zIndex: 1,
  },
  item: {
    height: 36,
    padding: '0 16px',
    display: 'flex',
    alignItems: 'center',
    fontSize: 12,
    color: '#585858',
    '&:hover': {
      backgroundColor: '#f4f6f7',
    },
  },
  checkMark: {
    marginLeft: 'auto',
  },
}

const DropdownFilterItemsList = ({
  classes,
  items,
  getItemProps,
  selectedItem,
}) => (
  <div className={classes.itemList}>
    {items.map(item => (
      <div {...getItemProps({ item, className: classes.item })} key={item.id}>
        {item.text}
        {selectedItem.id === item.id && (
          <SvgIcon
            width={9}
            color="#6e8be3"
            src={checkMark}
            className={classes.checkMark}
          />
        )}
      </div>
    ))}
  </div>
)

const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** Array of items to display */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
    }),
  ).isRequired,
  /** Function to get the props from Downshift */
  getItemProps: PropTypes.func.isRequired,
  /** The current selected item to be highlighted with the check mark  */
  selectedItem: PropTypes.shape({
    id: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
  }).isRequired,
  /** The width of the component */
  width: PropTypes.number.isRequired, // eslint-disable-line react/no-unused-prop-types
}

DropdownFilterItemsList.propTypes = propTypes

export default injectSheet(styles)(DropdownFilterItemsList)
