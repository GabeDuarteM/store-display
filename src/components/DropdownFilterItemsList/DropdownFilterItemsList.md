```js
<div style={{ height: 155 }}>
  <DropdownFilterItemsList
    items={[
      { id: 'price-asc', text: 'Price: low to high' },
      { id: 'price-desc', text: 'Price: high to low' },
      { id: 'rating-asc', text: 'Rating: low to high' },
      { id: 'rating-desc', text: 'Rating: high to low' },
    ]}
    getItemProps={customProps => customProps}
    selectedItem={{ id: 'price-asc', text: 'Price: low to high' }}
    width={200}
  />
</div>
```
