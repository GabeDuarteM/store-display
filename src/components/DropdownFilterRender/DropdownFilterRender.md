```js
<div style={{ height: 155 }}>
  <DropdownFilterRender
    label="Label"
    getLabelProps={customProps => customProps}
    getButtonProps={customProps => customProps}
    getItemProps={customProps => customProps}
    isOpen
    handleChangeMenuIsOpen={() => {}}
    selectedItem={{ id: 'price-asc', text: 'item 1' }}
    width={200}
    items={[
      { id: 'price-asc', text: 'item 1' },
      { id: 'price-desc', text: 'item 2' },
      { id: 'item-3', text: 'item 3' },
    ]}
  />
</div>
```
