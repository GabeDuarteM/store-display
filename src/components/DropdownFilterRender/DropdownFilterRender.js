import React from 'react'

import PropTypes from 'prop-types'
import injectSheet from 'react-jss'

import DropdownFilterInput from '../DropdownFilterInput'
import DropdownFilterItemsList from '../DropdownFilterItemsList'

export const styles = {
  renderRoot: {
    display: 'flex',
    alignItems: 'center',
    cursor: 'default',
    userSelect: 'none',
  },
  label: {
    fontSize: 12,
    color: '#585858',
    fontWeight: 500,
    marginRight: 8,
  },
}

const DropdownFilterRender = ({
  classes,
  label,
  getLabelProps,
  getButtonProps,
  getItemProps,
  isOpen,
  handleChangeMenuIsOpen,
  selectedItem,
  width,
  items,
}) => (
  <div className={classes.renderRoot}>
    {label && (
      <label className={classes.label} {...getLabelProps()}>
        {label}
      </label>
    )}
    <div>
      <DropdownFilterInput
        getButtonProps={getButtonProps}
        isOpen={isOpen}
        handleChangeMenuIsOpen={handleChangeMenuIsOpen}
        selectedItem={selectedItem}
        width={width}
      />
      {isOpen ? (
        <DropdownFilterItemsList
          items={items}
          getItemProps={getItemProps}
          selectedItem={selectedItem}
          width={width}
        />
      ) : null}
    </div>
  </div>
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** If provided, it displays a label to the dropdown */
  label: PropTypes.string,
  /** Function to get the props from Downshift */
  getLabelProps: PropTypes.func.isRequired,
  /** Function to get the props from Downshift */
  getButtonProps: PropTypes.func.isRequired,
  /** Function to get the props from Downshift */
  getItemProps: PropTypes.func.isRequired,
  /** Defines if the dropdown menu should be open */
  isOpen: PropTypes.bool.isRequired,
  /**
   * Gets called when the dropdown gets or loses focus
   * @param {bool} menuIsOpen The value that defines if the menu should or should not be open
   */
  handleChangeMenuIsOpen: PropTypes.func.isRequired,
  /** The current selected item to be displayed on the input */
  selectedItem: PropTypes.shape({
    id: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
  }).isRequired,
  /** The width of the component */
  width: PropTypes.number.isRequired,
  /** Array of items to display */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
    }),
  ).isRequired,
}

export const defaultProps = {
  label: null,
}

DropdownFilterRender.propTypes = propTypes
DropdownFilterRender.defaultProps = defaultProps

export default injectSheet(styles)(DropdownFilterRender)
