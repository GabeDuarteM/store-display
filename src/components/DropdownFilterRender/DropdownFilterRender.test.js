import React from 'react'

import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import DropdownFilterRender, {
  styles,
  propTypes,
  defaultProps,
} from './DropdownFilterRender'
import DropdownFilterInput from '../DropdownFilterInput'
import DropdownFilterItemsList from '../DropdownFilterItemsList'

const createWrapper = (func, optionalProps) =>
  func(
    <DropdownFilterRender
      getLabelProps={customProps => customProps}
      getButtonProps={customProps => customProps}
      getItemProps={customProps => customProps}
      isOpen
      handleChangeMenuIsOpen={() => {}}
      selectedItem={{ id: 'item-1', text: 'item 1' }}
      width={200}
      items={[
        { id: 'item-1', text: 'item 1' },
        { id: 'item-2', text: 'item 2' },
        { id: 'item-3', text: 'item 3' },
      ]}
      {...optionalProps}
    />,
  )

const createWrapperShallow = optionalProps =>
  createWrapper(shallow, optionalProps)

const createWrapperMount = optionalProps => createWrapper(mount, optionalProps)

describe('COMPONENT: <DropdownFilterRender />', () => {
  it('should render without crashing with default props', () => {
    createWrapperMount()
  })

  it('should render without crashing with custom props', () => {
    createWrapperMount({ label: 'Something' })
  })

  it("should match component's snapshot", () => {
    const wrapper = createWrapperShallow()
    expect(toJson(wrapper.dive())).toMatchSnapshot()
  })

  it("should match component's style snapshot", () => {
    expect(styles).toMatchSnapshot()
  })

  it("should match component's propTypes snapshot", () => {
    expect(propTypes).toMatchSnapshot()
  })

  it("should match component's defaultProps snapshot", () => {
    expect(defaultProps).toMatchSnapshot()
  })

  it('should display a label with the correct text if the prop is provided', () => {
    const wrapper = createWrapperShallow({ label: 'Something' })

    const label = wrapper.dive().find('label')
    const labelExists = label.exists()
    const labelText = label.text()

    expect(labelExists).toBeTruthy()
    expect(labelText).toBe('Something')
  })

  it('should not display a label if the prop is not provided', () => {
    const wrapper = createWrapperShallow()

    const label = wrapper.dive().find('label')
    const labelExists = label.exists()

    expect(labelExists).toBeFalsy()
  })

  it('should have a DropdownFilterInput component', () => {
    const wrapper = createWrapperShallow()

    const dropdownFilterInput = wrapper.dive().find(DropdownFilterInput)
    const dropdownFilterInputExists = dropdownFilterInput.exists()

    expect(dropdownFilterInputExists).toBeTruthy()
  })

  it('should have a DropdownFilterItemsList component if the isOpen prop is true', () => {
    const wrapper = createWrapperShallow({ isOpen: true })

    const dropdownFilterItemsList = wrapper.dive().find(DropdownFilterItemsList)
    const dropdownFilterItemsListExists = dropdownFilterItemsList.exists()

    expect(dropdownFilterItemsListExists).toBeTruthy()
  })

  it('should not have a DropdownFilterItemsList component if the isOpen prop is false', () => {
    const wrapper = createWrapperShallow({ isOpen: false })

    const dropdownFilterItemsList = wrapper.dive().find(DropdownFilterItemsList)
    const dropdownFilterItemsListExists = dropdownFilterItemsList.exists()

    expect(dropdownFilterItemsListExists).toBeFalsy()
  })
})
