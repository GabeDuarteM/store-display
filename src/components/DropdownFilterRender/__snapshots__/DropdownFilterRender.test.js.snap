// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT: <DropdownFilterRender /> should match component's defaultProps snapshot 1`] = `
Object {
  "label": null,
}
`;

exports[`COMPONENT: <DropdownFilterRender /> should match component's propTypes snapshot 1`] = `
Object {
  "classes": [Function],
  "getButtonProps": [Function],
  "getItemProps": [Function],
  "getLabelProps": [Function],
  "handleChangeMenuIsOpen": [Function],
  "isOpen": [Function],
  "items": [Function],
  "label": [Function],
  "selectedItem": [Function],
  "width": [Function],
}
`;

exports[`COMPONENT: <DropdownFilterRender /> should match component's snapshot 1`] = `
<div
  className="DropdownFilterRender-renderRoot"
>
  <div>
    <Jss(DropdownFilterInput)
      getButtonProps={[Function]}
      handleChangeMenuIsOpen={[Function]}
      isOpen={true}
      selectedItem={
        Object {
          "id": "item-1",
          "text": "item 1",
        }
      }
      width={200}
    />
    <Jss(DropdownFilterItemsList)
      getItemProps={[Function]}
      items={
        Array [
          Object {
            "id": "item-1",
            "text": "item 1",
          },
          Object {
            "id": "item-2",
            "text": "item 2",
          },
          Object {
            "id": "item-3",
            "text": "item 3",
          },
        ]
      }
      selectedItem={
        Object {
          "id": "item-1",
          "text": "item 1",
        }
      }
      width={200}
    />
  </div>
</div>
`;

exports[`COMPONENT: <DropdownFilterRender /> should match component's style snapshot 1`] = `
Object {
  "label": Object {
    "color": "#585858",
    "fontSize": 12,
    "fontWeight": 500,
    "marginRight": 8,
  },
  "renderRoot": Object {
    "alignItems": "center",
    "cursor": "default",
    "display": "flex",
    "userSelect": "none",
  },
}
`;
